sway-core/src/semantic_analysis/ast_node/expression/match_expression/analysis/constructor_factory.rs

```diff
@@ -278,9 +293,343 @@ impl ConstructorFactory {
+    /// elements of Σ.
+    ///
+    /// For example, a Σ composed of `Pattern::U64(..)`s would need to check for
+    /// if it is a complete signature for the `U64` pattern type. Versus a Σ
+    /// composed of `Pattern::Tuple([.., ..])` which would need to check for if
+    /// it is a complete signature for "`Tuple` with 2 sub-patterns" type.
+    ///
+    /// There are several rules with which to determine if Σ is a complete
+    /// signature:
+    ///
+    /// 1. If Σ is empty it is not a complete signature.
+    /// 2. If Σ contains only wildcard patterns, it is not a complete signature.
+    /// 3. If Σ contains all constructors for the type of the elements of Σ then
+    ///    it is a complete signature.
+    ///
+    /// For example, given this Σ:
+    ///
+    /// ```ignore
+    /// [
+    ///     Pattern::U64(Range { first: 0, last: 0 }),
+    ///     Pattern::U64(Range { first: 7, last: 7 })
+    /// ]
+    /// ```
+    ///
+    /// this would not be a complete signature as it does not contain all
+    /// elements from the `U64` type.
+    ///
+    /// Given this Σ:
+    ///
+    /// ```ignore
+    /// [
+    ///     Pattern::U64(Range { first: std::u64::MIN, last: std::u64::MAX })
+    /// ]
+    /// ```
+    ///
+    /// this would be a complete signature as it does contain all elements from
+    /// the `U64` type.
+    ///
+    /// Given this Σ:
+    ///
+    /// ```ignore
+    /// [
+    ///     Pattern::Tuple([
+    ///         Pattern::U64(Range { first: 0, last: 0 }),
+    ///         Pattern::Wildcard
+    ///     ]),
+    /// ]
+    /// ```
+    ///
+    /// this would also be a complete signature as it does contain all elements
+    /// from the "`Tuple` with 2 sub-patterns" type.
+    pub(crate) fn is_complete_signature(
+        &self,
+        pat_stack: &PatStack,
+        span: &Span,
+    ) -> CompileResult<bool> {
+        let mut warnings = vec![];
+        let mut errors = vec![];
+        let preprocessed = pat_stack.flatten().filter_out_wildcards();
+        if preprocessed.is_empty() {
+            return ok(false, warnings, errors);
+        }
+        let (first, rest) = check!(
+            preprocessed.split_first(span),
+            return err(warnings, errors),
+            warnings,
+            errors
+        );
+        match first {
+            // its assumed that no one is ever going to list every string
+            Pattern::String(_) => ok(false, warnings, errors),
+            // its assumed that no one is ever going to list every B256
+            Pattern::B256(_) => ok(false, warnings, errors),
+            Pattern::U8(range) => {
+                let mut ranges = vec![range];
+                for pat in rest.into_iter() {
+                    match pat {
+                        Pattern::U8(range) => ranges.push(range),
+                        _ => {
+                            errors.push(CompileError::Internal("type mismatch", span.clone()));
```

Comment:
This would be a failure of one of the invariants of the match exhaustivity algorithm---the invariant basically says that any time this function is call from inside the algorithm, that we should expect all of the patterns that are passed in to be similar. There are situations that would cause this, but they are caught before this algorithm is run and are considered `TypeError`'s. I suppose that I can't be certain that its not possible for some combination of bad to trigger this though. I'll make the error message a little more clear.